### ISSUE TITLE
Re-screen approved providers monthly

### ISSUE THREAD

**Issue by cecilia-donnelly** [SolutionGuidance/psm#740]
>The PSM shall support admin-configurable automated re-screening.  C.f. psm-FR-7.3.

And possibly:
>psm-FR-6.1 The PSM shall download all monitoring risk scores for each month as a CSV

The way I understand this is that the PSM will re-run automated checks on all approved applications on a monthly basis (though we should make the exact period configurable in Drools).  Right now, the only automated check is the LEIE, but by the end of June we'll have another one.

This is another item that we could use some more input about: how will users use this re-screening information?  What's an appropriate UI for this?  The CSV?  Notification emails about new exclusions sent to designated users?

**Comment by jasonaowen** [comment]
The next step that I'm working on (being the next-easiest thing, I think) is to allow admin reviewers to manually queue an automatic re-screening. This will bring in the `cause` and `created_by` attributes from the schema plan described in #787, as there will now be more than one reason for an automatic screening to have been created.

One aspect I'm thinking through is how to concurrently run the screening task. Up until now, we've only been running the LEIE check via a step in the jBPM business process; soon, we'll have more contexts to consider! For the admin-requested re-screening, we'll need to run the task outside of the web request handler thread, so that the user is not blocked on this background task - we don't want any of the things that can go wrong in an API request to result in a request timeout, or worse; we want to queue the work and then return a result saying we did so. Later, we'll also need to run a scheduled task.

So, to recap, three contexts the LEIE (and eventually other external screenings) needs to be run/queued in:

- jBPM, for new enrollment applications
- web request, for admin-requested re-screenings
- scheduled task, for monthly (or whatever period) re-screenings

There are a few different options I can think of for handling these:

- run in the current thread
  - will be the strategy for jBPM
  - not acceptable for web request
  - good for scheduled tasks
- run as jBPM business process: create a new single-node process (and add to it later for other external screeners) and rely on jBPM to run the screenings in the background
  - not okay for web requests or scheduled tasks; I'm trying to remove jBPM, not increase our dependence on it!
- run as event-driven process: set up a [message-driven bean](https://en.wikipedia.org/wiki/Enterprise_JavaBeans#Message_driven_beans), whose execution is automatically scheduled by the container
  - this is the original replacement to jBPM I was envisioning
  - web request would create event and then return
  - scheduled task would create event
  - somewhat high overhead: users will need to set up a message queue (which we [once had](https://github.com/SolutionGuidance/psm/pull/220)!)
  - how closely do we want to replicate the current jBPM business process? instead, maybe we can replace it with something simpler and easier to understand
- run as a [Spring `TaskExecutor`](https://docs.spring.io/spring/docs/4.3.12.RELEASE/spring-framework-reference/html/scheduling.html): let Spring interact with the container's managed thread pools, and/or spin off a new thread in a container-safe way
  - jBPM would continue to run as it does now
  - web request would queue task and return
  - scheduled task are very similar, and would probably be used for 
- bring in [RxJava](https://github.com/ReactiveX/RxJava) and depend on it to schedule tasks in the background
  - might be able to encode our current jBPM workflow entirely in several lines of RxJava code?
  - web request would queue task and return
  - it looks really neat! (woo resume-driven development)
  - this is probably not a dependency to bring in lightly

Right now, I'm leaning towards using a `TaskExecutor` for admin-requested re-screenings. They seem to be pretty straightforward, they don't need any additional dependencies, and of our options, I think they're the simplest possible thing. If they turn out to be a disaster for some reason, I also don't think it will be too big a deal to move away from them.

**Comment by jasonaowen** [comment]
@cecilia-donnelly, @frankduncan, @jvasile, @mwolfthal, and I have discussed how to handle the background execution of admin-requested re-screenings further. The two main options we considered were message queues and `TaskExecutor`.

Using a message-driven system has some advantages and some disadvantages. It provides a natural place to add audit logging, by allowing us to add additional message listeners. That also allows us to build around more of an audit-logging / event-sourcing mental model, by sending messages that record "this just happened" and let relevant components react to that new information; for sophisticated deployments, this may even include components external to the PSM.

This flexibility and separation comes at a cost, however, which is that it becomes more difficult to reason about the system. Tracing the link between "a fires an event" and "b responds to an event" is much more difficult than tracing the link between "a calls b"; the latter is very explicit in the code. It also requires additional configuration in the application server, which at this stage is already a bit of a challenging setup for new contributors.

Under either approach, we plan to have automatic screening audit logging in at least one representation: the [`TriggeringEvent` described in #787](https://github.com/SolutionGuidance/psm/pull/787), which captures why an automatic screening was run (new enrollment application, periodic rescreening, or admin clicked button, as well as when that trigger happened), and the `AutomaticScreening` result, which captures pass/fail/error status and any associated search results from the screening.

As I mentioned in the previous comment, for the purpose of queuing a re-screening in response to the admin pressing a button, I estimate that switching from `TaskExecutor` to message-driven beans to be inexpensive.

We also discussed error handling. Our current approach, which I don't intend to modify as part of this work, is to simply record failures: if an external service (such as our LEIE API) is unavailable for whatever reason, create a new `AutomaticScreening` with an `ERROR` result and be done. Admins can see and act on that accordingly, perhaps by pushing the button we're building to request a re-screening, or perhaps by manually checking with the external data source directly. Future work might include automatically re-running the screening on enrollment applications whose most recent screening resulted in an error.

Given all that, we decided that right now, for this particular feature, the complexity of a message-driven system is not worthwhile. If we learn of additional requirements, or have other features that turn out to be best implemented with a message-driven architecture, we'll happily revisit this. In the meantime, when the admin user presses the re-screen button, we'll use `TaskExecutor` to schedule that work in the background.

**Comment by jasonaowen** [comment]
I've split remaining related work off into a number of new issues:

- #924 Add audit logging to external screening
- #925 Allow admin users to trigger re-screening
- #926 Handle monthly re-screening errors
- #927 Send notification(s) on monthly re-screening failures

Those, along with now-merged PR #910 Re-screen approved enrollments monthly, I think resolve this issue.
